From 89cc9519a677be0208678a3e3c80584b4f7c7a74 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Mon, 3 Aug 2015 17:27:41 -0700
Subject: [PATCH 21/34] Compatibility fix for "ListenAddress ::"

In SunSSH, a config that specifies only "ListenAddress ::" in
fact will listen on both IPv4 and IPv6.
---
 servconf.c | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/servconf.c b/servconf.c
index 7e6c7f8..1b6cc34 100644
--- a/servconf.c
+++ b/servconf.c
@@ -769,11 +769,22 @@ process_queued_listen_addrs(ServerOptions *options)
 		options->address_family = AF_UNSPEC;
 
 	for (i = 0; i < options->num_queued_listens; i++) {
-		add_listen_addr(options, options->queued_listen_addrs[i],
-		    options->queued_listen_ports[i]);
+		/*
+		 * To retain backwards compat with SunSSH, convert any listener
+		 * on :: into a listener of unspecified AF (so it listens on
+		 * v4 too)
+		 */
+		if (strcmp(options->queued_listen_addrs[i], "::") == 0) {
+			add_listen_addr(options, NULL,
+			    options->queued_listen_ports[i]);
+		} else {
+			add_listen_addr(options, options->queued_listen_addrs[i],
+			    options->queued_listen_ports[i]);
+		}
 		free(options->queued_listen_addrs[i]);
 		options->queued_listen_addrs[i] = NULL;
 	}
+
 	free(options->queued_listen_addrs);
 	options->queued_listen_addrs = NULL;
 	free(options->queued_listen_ports);
-- 
2.5.4 (Apple Git-61)

